### PR TITLE
Make Kubernetes aware of the LoadBalancer behaviour

### PR DESCRIPTION
**Pull request by RyanAoh**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Implementation of [KEP-1860](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding)

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Added a new `ipMode` field to the `.status` of Services where `type` is set to `LoadBalancer`.
The new field is behind the `LoadBalancerIPMode` feature gate.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding
```

### CODE REVIEW

**Comment by RyanAoh** [comment]
A copy of https://github.com/kubernetes/kubernetes/pull/118895, since it was reverted for release resons.
BTW fix https://github.com/kubernetes/kubernetes/pull/118895#discussion_r1265664692

/milestone v1.29

**Comment by RyanAoh** [comment]
cc @thockin

**Comment by sftim** [comment]
Changelog suggestion:
```markdown
Added a new `ipMode` field to the `.status` of Services where `type` is set to `LoadBalancer`.
The new field is behind the `LoadBalancerIPMode` feature gate.
```

**Review by thockin — COMMENTED** [review]
I didn't see anything notably different from the previous PR, is that right?

/approve
/lgtm

File: pkg/proxy/service.go [hunk 1]
```diff
@@ -211,26 +207,22 @@ func (sct *ServiceChangeTracker) newBaseServiceInfo(port *v1.ServicePort, servic
...
-		ipFamilyMap = proxyutil.MapIPsByIPFamily(ips)
-
-		if ipList, ok := ipFamilyMap[proxyutil.OtherIPFamily(sct.ipFamily)]; ok && len(ipList) > 0 {
-			klog.V(4).InfoS("Service change tracker ignored the following load balancer ingress IPs for given Service as they don't match the IP Family",
-				"ipFamily", sct.ipFamily, "loadBalancerIngressIps", strings.Join(ipList, ", "), "service", klog.KObj(service))
-		}
-		// Create the LoadBalancerStatus with the filtered IPs
-		for _, ip := range ipFamilyMap[sct.ipFamily] {
-			info.loadBalancerStatus.Ingress = append(info.loadBalancerStatus.Ingress, v1.LoadBalancerIngress{IP: ip})
+		if ipFamily := proxyutil.GetIPFamilyFromIP(ing.IP); ipFamily == sct.ipFamily && proxyutil.IsVIPMode(ing) {
```
Comment (aojea) [inline pkg/proxy/service.go:216]: this need unit test coverage, added in https://github.com/kubernetes/kubernetes/pull/120068

File: pkg/proxy/service.go [hunk 2]
```diff
@@ -211,26 +207,22 @@ func (sct *ServiceChangeTracker) newBaseServiceInfo(port *v1.ServicePort, servic
...
+			info.loadBalancerVIPs = append(info.loadBalancerVIPs, ing.IP)
+		} else {
+			invalidIPs = append(invalidIPs, ing.IP)
```
Comment (aojea) [inline pkg/proxy/service.go:219]: we should not log IPs with ipMode proxy as invalid, fixed in https://github.com/kubernetes/kubernetes/pull/120068

**Pull request merged**